course5/homework02/homework0201.cc

```diff
@@ -0,0 +1,50 @@
+#include <iostream>
+#include <cmath>
+#include <cfloat>
+
+/****
+homework02, part 1
+Implement Carmack’s fast square root. Compare the values it produces with the output of std::sqrt for a range of values, draw conclusions.
+
+Carmack’s fast square root, from Wikipedia:
+an algorithm that estimates 1/sqrt(x), the reciprocal (or multiplicative inverse) of the square root of a 32-bit floating-point number x in IEEE 754 floating-point format.
+The algorithm accepts a 32-bit floating-point number as the input and stores a halved value for later use. Then, treating the bits representing the floating-point number as a 32-bit integer, a logical shift right by one bit is performed and the result subtracted from the magic number 0x5f3759df. This is the first approximation of the inverse square root of the input. Treating the bits again as a floating-point number, it runs one iteration of Newton's method, yielding a more precise approximation.
+***/
+
+float carmack_sqrt(float input)
+{
+    long* fl_to_int = reinterpret_cast<long*>(&input);
+
+    long shifted_rt = *fl_to_int >> 1;
+    long from_magic = 0x5f3759df - shifted_rt;
+
+    float* int_to_fl = reinterpret_cast<float*>(&from_magic);
```

Comment:
Class notes said "Very dangerous!" -- is there another way to do it? (I'll check out some other responses)